pyinfra/api/connectors/winrm.py

```diff
@@ -0,0 +1,188 @@
+        ('timeout', state.config.CONNECT_TIMEOUT),
+    ):
+        if value:
+            kwargs[key] = value
+
+    # Password auth (boo!)
+    if host.data.winrm_password:
+        kwargs['password'] = host.data.winrm_password
+
+    # TODO: add more auth
+    # pywinrm supports: basic, certificate, ntlm, kerberos, plaintext, ssl, credssp
+    # see https://github.com/diyan/pywinrm/blob/master/winrm/__init__.py#L12
+
+    return kwargs
+
+
+def connect(state, host):
+    '''
+    Connect to a single host. Returns the winrm Session if successful.
+    '''
+
+    kwargs = _make_winrm_kwargs(state, host)
+    logger.debug('Connecting to: {0} ({1})'.format(host.name, kwargs))
+
+    # Hostname can be provided via SSH config (alias), data, or the hosts name
+    hostname = kwargs.pop(
+        'hostname',
+        host.data.ssh_hostname or host.name,
+    )
+
+    logger.debug('winrm_username:{}'.format(host.data.winrm_username))
+    logger.debug('winrm_password:{}'.format(host.data.winrm_password))
+    logger.debug('winrm_port:{}'.format(host.data.winrm_port))
+    logger.debug('use_shell:{}'.format(host.data.use_shell))
+
+    try:
+        # Create new session
+        host_and_port = '{}:{}'.format(hostname, host.data.winrm_port)
+        logger.debug('host_and_port: {}'.format(host_and_port))
+
+        session = winrm.Session(
+            host_and_port,
+            auth=(
+                host.data.winrm_username,
+                host.data.winrm_password,
+            ),
+        )
+
+        return session
+
+    # TODO: add exceptions here
+    except Exception as e:
+        auth_kwargs = {}
+
+        for key, value in kwargs.items():
+            if key in ('username', 'password'):
+                auth_kwargs[key] = value
+                continue
+
+        auth_args = ', '.join(
+            '{0}={1}'.format(key, value)
+            for key, value in auth_kwargs.items()
+        )
+        logger.debug(str(e))
+        _raise_connect_error(host, 'Authentication error', auth_args)
+
+
+def run_shell_command(
+    state, host, command,
+    get_pty=False,
+    timeout=None,
+    stdin=None,
+    success_exit_codes=None,
+    print_output=False,
+    return_combined_output=False,
+    **command_kwargs
+):
+    '''
+    Execute a command on the specified host.
+
+    Args:
+        state (``pyinfra.api.State`` obj): state object for this command
+        hostname (string): hostname of the target
+        command (string): actual command to execute
+        get_pty (boolean): Not used for WINRM
+        env (dict): envrionment variables to set
+        timeout (int): timeout for this command to complete before erroring
+
+    Returns:
+        tuple: (exit_code, stdout, stderr)
+        stdout and stderr are both lists of strings from each buffer.
+    '''
+
+    #_shell_executable = 'cmd'
+    #if host.data.use_shell:
+        #_shell_executable = host.data.use_shell
+    command = make_command(command, **command_kwargs)
```

Comment:
`make_command` is designed for building Unix-like commands, I think this will need it's own command builder.

Sidenote: I'll rename `make_command` to `make_unix_command`.

Reply:
ok